src/components/core/PivotTable.tsx

```diff
@@ -0,0 +1,232 @@
+        super(props);
+        this.state = {
+            columnDefs: [],
+            rowData: [],
+            execution: null
+        };
+        this.gridDataSource = null;
+        this.gridApi = null;
+        visualizationIsBetaWarning();
+    }
+
+    public componentWillMount() {
+        const { resultSpec, getPage } = this.props;
+        this.createDataSource(resultSpec, getPage);
+    }
+
+    public componentWillReceiveProps(
+        nextProps: IPivotTableProps & ILoadingInjectedProps & IDataSourceProviderInjectedProps
+    ) {
+        const { resultSpec, getPage } = this.props;
+        if (
+            !isEqual(resultSpec, nextProps.resultSpec)
+            || getPage !== nextProps.getPage
+        ) {
+            this.createDataSource(resultSpec, getPage);
+            this.setGridDataSource();
+        }
+    }
+
+    public createDataSource = (resultSpec: AFM.IResultSpec, getPage: IGetPage) => {
+        const onSuccess = (execution: Execution.IExecutionResponses, columnDefs: IGridHeader[]) => {
+            this.setState({ execution, columnDefs });
+        };
+        this.gridDataSource = getGridDataSource(resultSpec, getPage, onSuccess);
+    }
+
+    public onGridReady = (params: IGridParams) => {
+        this.gridApi = params.api;
+        this.setGridDataSource();
+    }
+
+    public setGridDataSource = () => {
+        this.gridApi.setDatasource(this.gridDataSource);
+    }
+
+    public renderVisualization() {
+        const { columnDefs, rowData } = this.state;
+        const { pageSize } = this.props;
+
+        const gridOptions = {
+
+            // Initial data
+            columnDefs,
+            rowData,
+
+            // Basic options
+            suppressMovableColumns: true,
+            enableSorting: false,
+            enableFilter: false,
+            enableColResize: true,
+
+            // infinite scrolling model
+            rowModelType: 'infinite',
+            paginationPageSize: pageSize,
+            cacheOverflowSize: pageSize,
+            cacheBlockSize: pageSize,
+            maxConcurrentDatasourceRequests: 1,
+            infiniteInitialRowCount: pageSize,
+            maxBlocksInCache: 10,
+            onGridReady: this.onGridReady,
+
+            // Custom renderers
+            frameworkComponents: {
+                // loading indicator
+                // any is needed here because of incompatible types with AgGridReact types
+                loadingRenderer: null as any
+            }
+        };
+        const RowLoadingElement = (props: ILoadingRendererProps) =>
+            (props.node.id !== undefined
+                ? <span>{props.data[props.colDef.field]}</span>
+                : <LoadingComponent width={36} imageHeight={8} height={26} speed={2} />);
+
+        // this cannot be set directly in gridOptions because of incompatible types with AgGridReact types
+        gridOptions.frameworkComponents.loadingRenderer = RowLoadingElement;
```

Comment:
it can be written simpler like this:

```
            frameworkComponents: {
                // loading indicator
                // any is needed here because of incompatible types with AgGridReact types
                loadingRenderer: RowLoadingElement as any
            }
```

Reply:
that worked